ruby/bps/lib/bps/publisher/abstract.rb

```diff
@@ -11,6 +11,10 @@ def publish(_message, **_opts)
         def flush(**); end
       end
 
+      def initialize
+        ObjectSpace.define_finalizer(self, proc { close })
```

Comment:
@jpmedforth I added this to make sure the publisher is auto-closed but that only works if the producer is GC'd, which won't work if you assign it as a class method.